lib/geoengineer/gps/gps.rb

```diff
@@ -0,0 +1,305 @@
+require 'yaml'
+require 'pry'
+###
+# GPS Geo Planning System
+# This module is designed as a higher abstration above the "resources" level
+# The abstration is built for engineers to use so is in their vocabulary
+# As each engineering team is different GPS is only building blocks
+# GPS is not a complete solution
+###
+class GeoEngineer::GPS
+  class NotFoundError < StandardError; end
+  class NotUniqueError < StandardError; end
+  class BadQueryError < StandardError; end
+  class GPSProjetNotFound < StandardError; end
+  class NodeTypeNotFound < StandardError; end
+  class MetaNodeError < StandardError; end
+
+  ###
+  # HASH METHODS
+  ###
+
+  # remove_ removes all keys starting with `_`
+  def self.remove_(hash)
+    hash = hash.dup
+    hash.each_pair do |key, value|
+      hash.delete(key) && next if key.to_s.start_with?("_")
+      hash[key] = remove_(value) if value.is_a?(Hash)
+    end
+    hash
+  end
+
+  def self.deep_dup(object)
+    JSON.parse(object.to_json)
+  end
+
+  ###
+  # END OF HASH METHODS
+  ###
+
+  ###
+  # Search Methods
+  ###
+
+  # where returns multiple nodes
+  def self.where(nodes, query = "*:*:*:*:*")
+    search(nodes, query)
+  end
+
+  # find a node from nodes
+  def self.find(nodes, query = "*:*:*:*:*")
+    query_nodes = search(nodes, query)
+    raise NotFoundError, "for query #{query}" if query_nodes.empty?
+    raise NotUniqueError, "for query #{query}" if query_nodes.length > 1
+    query_nodes.first
+  end
+
+  def self.split_query(query)
+    query_parts = query.split(":")
+    raise BadQueryError, "for query #{query}" if query_parts.length != 5
```

Comment:
It _might_ be cool if you could also pass in an attribute, i.e. terraform_id -> `infra/codeflow:::api:terraform_id`

Reply:
This might be a useful extension later.